### PR TITLE
Document that writes to padding are not preserved

### PR DESCRIPTION
**Pull request by joshlf**
Document that writes to padding bytes are not preserved when an object is moved or copied.

Makes progress on #1648

### CODE REVIEW

**Comment by joshlf** [comment]
@saethlin Following up on our conversation at RustConf, does the wording in this warning accurately describe the behavior wrt padding bytes?

File: src/lib.rs
Comment (saethlin) [inline src/lib.rs:1958]: Rust does not have objects. We have allocations and typed or untyped copies. Untyped copies copy all bytes in the specified range, and currently the only way you can do this is `ptr::copy` and `ptr::copy_nonoverlapping`.

Typed copies are every other copy, which includes (some of these may be surprising to users) assignment, reads/writes through pointers, function arguments, and function return values.

Typed copies are not guaranteed to preserve bytes which are padding in all variants. That's probably what this comment means to identify.

Comment (joshlf) [reply src/lib.rs:1958]: Yeah, "typed copy" is what I'm referring to. Do you know if that term is defined anywhere in the Reference or stdlib docs? A quick Google doesn't turn anything up.

Comment (saethlin) [reply src/lib.rs:1958]: The docs for ptr::copy contain a probably-unhelpful mention of "untyped".

Comment (joshlf) [reply src/lib.rs:1958]: Do you know of any internal docs/discussions that define the term? Or is this just one of those "compiler people use these terms and know what they mean when they talk to each other and they weren't expecting you to look inside the compiler" things?

Comment (saethlin) [reply src/lib.rs:1958]: Ugh of course that doesn't work. 
@CAD97 @digama0 @JakobDegen @RalfJung do we have documentation for what a typed copy is?

Comment (RalfJung) [reply src/lib.rs:1958]: I don't know of any proper docs for this, no. That's clearly something we should fix. Any volunteers? :)

One question is where these docs would even go. The primary syntax for a typed copy is `a = b;`, so maybe this should be [here](https://doc.rust-lang.org/nightly/reference/expressions/operator-expr.html#assignment-expressions)?

Comment (jswrenn) [inline src/lib.rs:1958]:
```suggestion
/// Note that, when an value is moved or copied, only the non-padding bytes of
```

**Review by jswrenn — APPROVED** [review]
Approved, contingent on this change.

File: src/lib.rs
Comment (joshlf) [reply src/lib.rs]: Done

Comment (joshlf) [reply src/lib.rs]: [s/object/value](https://github.com/google/zerocopy/pull/1660#discussion_r1761911299), but otherwise I'm gonna merge this and we can nitpick the wording later once docs have landed to Rust that we can cite.

**Pull request merged**
